cli/azd/pkg/oneauth/oneauth_windows.go

```diff
@@ -0,0 +1,230 @@
+	int expiresOn;
+	char *token;
+} WrappedAuthResult;
+*/
+import "C"
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+	"unsafe"
+
+	"github.com/azure/azure-dev/cli/azd/internal"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"golang.org/x/sys/windows"
+)
+
+// Supported indicates whether this build includes OneAuth integration.
+const Supported = true
+
+var (
+	//go:embed bridge/_build/Release/bridge.dll
+	bridgeDLL []byte
+	//go:embed bridge/_build/Release/bridge.dll.sha256
+	bridgeChecksum string
+	//go:embed bridge/_build/Release/fmt.dll
+	fmtDLL []byte
+	//go:embed bridge/_build/Release/fmt.dll.sha256
+	fmtChecksum string
+
+	// bridge provides access to the OneAuth API
+	bridge       *windows.DLL
+	authenticate *windows.Proc
+	freeAR       *windows.Proc
+	logout       *windows.Proc
+	shutdown     *windows.Proc
+	startup      *windows.Proc
+)
+
+type authResult struct {
+	homeAccountID string
+	token         azcore.AccessToken
+}
+
+type credential struct {
+	authority     string
+	clientID      string
+	homeAccountID string
+	opts          CredentialOptions
+}
+
+// NewCredential creates a new credential that acquires tokens via OneAuth.
+func NewCredential(authority, clientID string, opts CredentialOptions) (azcore.TokenCredential, error) {
+	cred := &credential{
+		authority:     authority,
+		clientID:      clientID,
+		homeAccountID: opts.HomeAccountID,
+		opts:          opts,
+	}
+	return cred, nil
+}
+
+func (c *credential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	ar, err := authn(c.authority, c.clientID, c.homeAccountID, strings.Join(opts.Scopes, " "), c.opts.NoPrompt, c.opts.Debug)
+	if err == nil {
+		c.homeAccountID = ar.homeAccountID
+	}
+	return ar.token, err
+}
+
+func LogIn(authority, clientID, scope string, debug bool) (string, error) {
+	ar, err := authn(authority, clientID, "", scope, false, debug)
+	return ar.homeAccountID, err
+}
+
+func Logout(clientID string, debug bool) error {
+	err := start(clientID, debug)
+	if err == nil {
+		logout.Call()
+	}
+	return err
+}
+
+func start(clientID string, debug bool) error {
+	if started.CompareAndSwap(false, true) {
+		err := loadDLL()
+		if err != nil {
+			return err
+		}
+		dbg := uintptr(0)
+		if debug {
+			dbg = uintptr(1)
+		}
+		clientID := unsafe.Pointer(C.CString(clientID))
+		defer C.free(clientID)
+		appID := unsafe.Pointer(C.CString(applicationID))
+		defer C.free(appID)
+		v := unsafe.Pointer(C.CString(internal.VersionInfo().Version.String()))
+		defer C.free(v)
+		p, _, _ := startup.Call(uintptr(clientID), uintptr(appID), uintptr(v), dbg)
+		// startup returns a char* message when it fails
+		if p != 0 {
+			// reset started so the next call will try to start OneAuth again
+			started.CompareAndSwap(true, false)
+			// TODO: in principle we should free the pointer but doing so from Go can cause a crash
```

Comment:
That sounds terrifying - do you know why?  Is it because the DLL may end up being unloaded or something?

Reply:
No, this one's a mystery to me. In native code I use `strdup` to copy strings to the heap before returning (pointers to) them to Go, to avoid losing them when the native stack unwinds. Freeing these pointers from Go (i.e. calling `C.free`) causes heap corruption and access violations, however doing the same from C works as expected. I haven't found documentation to explain this. I left a TODO because the leak isn't a problem--it's a handful of bytes leaked when authentication fails i.e., right before azd exits--but it does sound scary. I've fixed the issue by wrapping this message in a struct and freeing that on the native side.